lib/matplotlib/layout_engine.py

```diff
@@ -0,0 +1,207 @@
+"""
+A module to define layout engines for Matplotlib.
+
+Figures have a ``layout_engine`` property that holds a subclass of
+`~.LayoutEngine` defined here (or *None* for no layout).  At draw time
+``figure.layout_engine.execute()`` is called, the goal of which is usually
+to rearrange axes on the figure to produce a pleasing layout.  Note that
+this could also be implimented as a ``draw`` callback, however when
+printing we often disable the layout engine for the final draw, so it is
+helpful to have a different callback mechanism. It is also useful to
+know the layout engine while the figure is being created, in particular
+to deal with colorbars, so having a dedicated property of the figure to
+querry is useful.
+
+Matplotlib has two native layout engines, ``tight_layout`` and
+``constrained_layout``, which are implimented using this formalism
+as `.tight_layout_engine` and `.constrained_layout_engine`.  While layout
+engines tend to be complicated, users and downstream libraries can now create
+their own layout engine and connect it to a figure.
+"""
+
+import matplotlib as mpl
+import matplotlib._api as _api
+
+from matplotlib._constrained_layout import do_constrained_layout
+from matplotlib.tight_layout import (get_subplotspec_list,
+                                     get_tight_layout_figure)
+from matplotlib.backend_bases import _get_renderer
+from contextlib import nullcontext
+
+
+class LayoutEngine():
+    """
+    Base class for matplotlib layout engines.
+
+    A layout engine can be passed to a figure at instantiation or
+    at any time with `~.figure.Figure.set_layout_engine`.  However, note
+    note that layout engines affect the creation of colorbars, so
+    `~.figure.Figure.set_layout_engine` should be called before any
+    colorbars are created.
+
+    Once attached to a figure, the layout engine ``execute`` function
+    is called at draw time by `~.figure.Figure.draw`, providing a special
+    draw-time hook.
+    """
+    def __init__(self, fig, **kwargs):
```

Comment:
Should `fig` be a parameter, or should we have a "figureless" approach where the engine doesn't know the figure, *except* when it gets passed in to `layout.execute(fig)`?
Note that the API you proposed at the top is
```
fig = plt.figure(layout=my_favourite_layout(option1=0.4, option2=0.7))
```
whereas if you have to pass in the figure, you would have to write
```
fig = plt.figure()
fig.set_layout_engine(my_favourite_layout(fig=fig, ...))
```
which is quite less practical.